_inc/client/components/upgrade-link/index.jsx

```diff
@@ -0,0 +1,72 @@
+/**
+ * External dependencies
+ */
+import React, { PureComponent } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import store from 'state/redux-store';
+
+/**
+ * Internal dependencies
+ */
+import { getSiteRawUrl, getAffiliateCode } from 'state/initial-state';
+
+/**
+ * Build the upgrade URL
+ *
+ * @param {string} source        Context where this URL is clicked.
+ * @param {string} siteRawUrl    RAW Url for this site.
+ * @param {string} affiliateCode The affiliate code.
+ *
+ * @return {string} Upgrade URL with source, site, and affiliate code added.
+ */
+const buildUpgradeUrl = ( source, siteRawUrl, affiliateCode ) => `https://jetpack.com/redirect/?${ [
+	`source=${ source }`,
+	`site=${ siteRawUrl }`,
+	...( '' === affiliateCode ? [] : [ `aff=${ affiliateCode }` ] ),
+].join( '&' ) }`;
+
+/**
+ * Return an upgrade URL
+ *
+ * @param {string} source Context where this URL is clicked.
+ *
+ * @return {string} Upgrade URL with source, site, and affiliate code added.
+ */
+export const getUpgradeUrl = source => {
```

Comment:
This function is just a selector. It should live somewhere inside `client/state`. Some reducer file. `state/initial-state/reducer.js` should be ok